### PR TITLE
Druid

### PR DESCRIPTION
**Pull request by julianhyde**
This is complementary to https://issues.apache.org/jira/browse/CALCITE-1121, Calcite's Druid adapter.

I've added a Druid installation and two Druid data sets (wikiticker, and a denormalized foodmart). Druid consists of no fewer than 5 JVMs, so takes quite a bit of memory. I had to bump up the VM's memory to 3GB. I really tried to go lower, but there is a significant chance that Druid indexing will crash.

If people just don't have enough memory, they can change the 3072 in Vagrantfile back to 1024. Druid will crash, but the other databases will probably carry on working.

Please review this pull request, and if you give a +1 I will squash and re-submit.

### CODE REVIEW

**Comment by vlsi** [comment]
I wonder if we are going to squeeze all the kinds of DBs into a single VM.
Well, 3G is not that big, however just adding stuff in does not look that scalable

**Comment by julianhyde** [comment]
Well, I thought it was big! I have a Macbook Air with 4G RAM and I wasted a few hours trying to get the VM back to 1G memory.

I don't think many of the systems we would have adapters for would require more than 256M memory in a default configuration. So we can probably fit a dozen in a 4G VM.

If a single VM becomes unmanageable we could parameterize `Vagrantfile`, enabling say ["mysql", "postgresql", "druid"] in one VM, and ["cassandra", "splunk"] in another. I don't know whether vagrant allows parameterization, but if not, we could have two directories, each of which has a `Vagrantfile` that is a subset of the main `Vagrantfile`.

**Comment by vlsi** [comment]
> but if not, we could have two directories, each of which has a Vagrantfile

I think in the same direction.

OK. Let's try to fit all in one and see how it goes.

**Comment by michaelmior** [comment]
This would be taking things in a very different direction but one alternative would be to isolate each database individually. I think this would likely be much easier to manage using Docker rather Vagrant. 

Each DB would have its own Dockerfile. This would make it easy for everyone to run only the services they need. In addition, if we publish the Docker images then it would avoid the need for everyone who wants to use it to build an image as in the case now with Vagrant. 

Thoughts?

**Comment by vlsi** [comment]
Last time we checked that Docker was not that good in OS X / Windows environments.

Have it become any better?

**Comment by michaelmior** [comment]
@vlsi I think it runs quite well these days. The Docker team has done a lot of work on making this experience better. Instead of [boot2docker](http://boot2docker.io/), there is now [Docker Machine](https://docs.docker.com/machine/overview/) for creating a lightweight Linux VM on OS X/Windows for running the containers.

**Comment by michaelmior** [comment]
Re: memory reduction. I bumped the limit up when I added Cassandra because the VM kept running out of memory. I suspect this will happen again now.

**Comment by vlsi** [comment]
Re docker:
1) Need to check UX.  Docker Machine alpha did not work for me somehow (OS X).
2) Does it cache downloaded artifacts (e.g. OS-inside-docker-vm updates)?

> Re: memory reduction

As far as I understand, Docker is not yet natively supported in OS X / Windows, thus it has to run in some host VM, thus it will hit similar memory limits. Will it?

**Comment by michaelmior** [comment]
1) Sure, doesn't hurt to take a closer look at UX if a lot of devs are using Windows/Mac. My experience was that this works pretty nicely.
2) Any layers only need to get pulled once regardless of being in the VM or native Linux.

Fair point about the memory issue. There does still need to be some limit set for the VM on OS X and Windows. However, I think making it so that it won't be necessary to set up all the database services still gives this a net win. Worst case scenario when someone actually needs every DB, it should be roughly on par with the current approach.

**Comment by vlsi** [comment]
> 2) Any layers only need to get pulled once regardless of being in the VM or native Linux.

In "boot2docker" times, layers were cached _inside_ boot2docker VM, thus you had to re-download all the stuff if you somehow recreated boot2docker VM.

**Comment by michaelmior** [comment]
@vlsi That may still be the case but I don't think it's a huge concern. Right now if you destroy the Vagrant VM, you have to redownload anything installed in the VM. So at worst still on par with the current situation.

**Comment by vlsi** [comment]
I've checked current docker-toolbox, and it works fine in OS X.
I'm fine with migrating to docker. Well, docker was my initial plan, however OS X support was bad.

AFAIK, Apache Infra allows docker in Apache Jenkins CI, so having docker could help there as well.

> Right now if you destroy the Vagrant VM, you have to redownload anything installed in the VM

Technically you are not that right.
`calcite-test-dataset/vm/target/apt` holds apt cache, thus *.deb files are reused.

**Comment by vlsi** [comment]
```
[INFO] parent ............................................. SUCCESS [  0.366 s]
[INFO] foodmart-dataset ................................... SUCCESS [ 14.785 s]
[INFO] zips ............................................... SUCCESS [  0.014 s]
[INFO] vm ................................................. SUCCESS [16:55 min]
[INFO] h2 ................................................. SUCCESS [  0.718 s]
[INFO] hsqldb ............................................. SUCCESS [ 25.056 s]
[INFO] mysql .............................................. SUCCESS [03:14 min]
[INFO] postgresql ......................................... SUCCESS [  0.167 s]
[INFO] twissandra ......................................... SUCCESS [  0.043 s]
[INFO] test-dataset ....................................... SUCCESS [  0.005 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
```

**Comment by julianhyde** [comment]
@vlsi Thanks for merging! But you didn't give me chance to squash the commits. Can you please squash the commits into one and force push?

**Comment by michaelmior** [comment]
As an aside, it took over an hour to provision the VM with this change versus a few minutes last time I did so. I didn't look closely at the changes but is there any way to get the runtime down?

**Comment by julianhyde** [comment]
On my machine (a 4 yr old quadcore linux desktop with 24GB RAM) it took 6m 8s before, 13m 26s after. So, it's gotten slower, but not as bad as you see it. Are you running out of real memory and swapping?

We could maybe speed up the process by running init_druid.sh in parallel with the other init_ scripts. But that won't help much if you're short of real memory.

**Comment by michaelmior** [comment]
My machine has 6 cores and 16GB RAM with around 3GB free. I haven't looked at why it seems to be taking so long. I see a bunch of garbled output that looks like something trying to output a progress bar to a TTY. Perhaps for downloading something? That said, my connection here is quite fast so I wouldn't expect that to be a bottleneck either. In case, I'd personally like to see issues like this tackled via #8.

**Comment by julianhyde** [comment]
Yes, the first time you run it it will download druid-0.9.0-bin.tar.gz, about 130MB. After that it is cached in vm/target/apt so it should run faster.

File: foodmart-dataset/src/main/resources/druid/run.sh [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    io.druid.cli.Main server coordinator 2>&1 >coordinator.log &
+sleep 30
+
+echo Start Druid overlord node
+java `conf overlord | xargs` \
+    -cp conf-quickstart/druid/_common:conf-quickstart/druid/overlord:lib/* \
+    io.druid.cli.Main server overlord 2>&1 >overlord.log &
+sleep 30
+
+echo Start Druid middle manager node
```
Comment (b-slim) [inline foodmart-dataset/src/main/resources/druid/run.sh:52]: i know this PR is merged, but if you want to save same mem space within the VM, you can avoid having a middle manager. In Druid Middle manager are the same as overlord and they are used for horizontal scalability.

Comment (julianhyde) [inline foodmart-dataset/src/main/resources/druid/run.sh:52]: Thanks for the tip. I have logged https://github.com/vlsi/calcite-test-dataset/issues/9.

**Pull request merged**